### ISSUE TITLE
Implement "Variational Autoencoder" with ADVI

### ISSUE THREAD

**Issue by pboes** [TuringLang/Turing.jl#1885]
Is it possible, using the VI capabilities of Turing, to train the "ADVI analogue" of a Variational Autoencoder using Turing?

By that I mean the following: While it is of course possible to implement a VAE in Pytorch, one can also conveniently do so in Pyro (see https://pyro.ai/examples/vae.html), using the package's syntax to specify both a model and a "guide" (aka parametrized family of approximate posteriors). One then loads both of these into the pyro's SVI class which takes care of the rest.

By analogy, while it is possible to implement a VAE in Flux, I'm wondering whether it is possible to the same using Turing's syntax. To be precise, when I say "the same", I don't necessarily mean that one has to be able to specify an encoder network as a "guide", but simply to train a generative deep learning model $P_\theta(X|Z)P(Z)$ using variational inference. 

Thinking about this, my first idea was: Use ADVI to get approximate posteriors, then update the weights of the decoder network using samples from the posterior and a suitable loss function, and repeat. However, the magic of VAEs is that the posterior and the generative model are trained simultaneously, whereas in this approach, I would have to run Turing's `Variational.vi` once per packprop step, so that doesn't seem scalable.

The other idea would be to use ADVI in an EM algorithm, but that really doesn't have much to do with the original idea anymore...Anyway, any help on this would be appreciated.

**Comment by Red-Portal** [comment]
Hi, this itself is not easy to do using Turing not only in terms of scalability and performance, but also library design. The thing is that the VAE objective uses the closed-form KL divergence, which is a very specific case. The best thing do to would be to implement things using Flux as in this [example](https://github.com/FluxML/model-zoo/tree/master/vision/vae_mnist)
